simulator/infra/macro.h

```diff
@@ -218,4 +218,39 @@ auto test_subtraction_overflow( T_src1 src1, T_src2 src2)
     return std::make_pair( narrow_cast<T>( result), is_overflow);
 }
 
+template<typename T> static
+auto shuffle_stage( T src, uint64_t maskL, uint64_t maskR, int N)
+{
+    auto x = src & ~( maskL | maskR);
+    x |= ( ( src << N) & maskL) | ( ( src >> N) & maskR);
+    return x;
+}
+
+// array with constants for shuffle
+const uint64_t shfl_consts[] = {
```

Comment:
We do not use C-style arrays, please use C++ `std::array`.
Also, it might violate ODR.